Expose run_async function on torch::jit::Method

ShijunK opened the pull request:
Summary:
This change builds on top of
https://github.com/pytorch/pytorch/pull/46865

further exposing the async interface to `torch::jit::Method`.

Test Plan: `buck test caffe2/test/cpp/jit/...`

Differential Revision: D25219726

houseroad reviewed (commented):
Can we add a unittest?

suo commented:
I'm having second thoughts about this general approach…can you explain why it isn't sufficient to just express the asynchrony in the user code (using fork/wait)? Why does there need to be an explicit `runAsync` on `Method`?

ShijunK commented:
> I'm having second thoughts about this general approach…can you explain why it isn't sufficient to just express the asynchrony in the user code (using fork/wait)? Why does there need to be an explicit `runAsync` on `Method`?

Sure, this is a follow up of https://github.com/pytorch/pytorch/pull/46865, which exposes async all the way to graph executor. However, until this PR, there is no hook/public interface for user to inject a customized task launcher (probably a threadpool backed executor). Of cause, injecting customized task launcher not necessary means we have to `runAsync`, it is perfect fine to run model in caller's thread until hitting async ops (like fork/rpc) then switch to use passed in task launcher for async ops and wait on caller thread. The `runAsync` on method provides a choice for user to also move non-async part of model inference to given task launcher and immediately return a future. One use case could be, IO thread pool of a server handlding incoming model inference requests, using `runAsync` to offload entire inference workload to a separate cpu thread pool.

ShijunK commented:
> Can we add a unittest?

added one in `test/cpp/jit/test_module_api.cpp`, leverage the saved model which used for graph executor async test.

dzhulgakov commented on torch/csrc/jit/api/method.h:
```diff
@@ -31,6 +31,11 @@ struct TORCH_API Method {
       std::vector<c10::IValue> stack,
       const Kwargs& kwargs = Kwargs());
 
+  c10::intrusive_ptr<c10::ivalue::Future> run_async(
```
add a comment talking about threading model. What runs on the caller thread, when the return happens, etc.

dzhulgakov reviewed (commented):
@suo - in short, you usually want to propagate async APIs all the way up to the stack to avoid caller thread to block. If the model is compute-heavy and uses only 'transparent' thread pools, indeed the sync API should be enough and we could instead make "caller" thread to participate in the pool for the duration of method execution (though we don't do it now :( ).

However, there are cases where async ops (i.e. ops returning futures) wrap opaque thread pool or do some IO work. In this case exposing async entry point to TorchScript makes sense and allows to build more efficient service wrappers

dzhulgakov commented on torch/csrc/jit/api/module.cpp:
```diff
@@ -118,6 +118,17 @@ IValue Method::operator()(std::vector<IValue> stack, const Kwargs& kwargs) {
+    const Kwargs& kwargs,
+    TaskLauncher taskLauncher) {
+  stack.insert(stack.begin(), owner()._ivalue());
+  RECORD_TORCHSCRIPT_FUNCTION(name(), stack);
```
@iliacher to see whether the logging keeps working as expected (because the destructor of RecordFunction will be called before the function really finishes)

ilia-cher replied:
yep it won't work this way - you'd only log the async launch, not the method execution itself

dzhulgakov replied:
any suggestions on how to fix it? I guess we need to construct RecordFunction as unique_ptr and destroy it only when the future is satisfied. Will it work if it happens on a different thread?

alternatively if we keep it as-is, would it be a problem for profiling? I guess it might cause the event ranges not perfectly nested

I guess in general we should have a coherent story how profiler ranges should work with multithreading

ilia-cher replied:
yes, RecordFunction is ok with finishing on another thread, but can we instead just trace the actual async job? if you absolutely have to trace the whole [launch, end of async job] span then you could move Recordfunction into the async lambda (e.g. async job, or a callback attached to async job future)

ilia-cher replied:
(by move I mean, create e.g. a unique_ptr in launch section and move it into the lambda)

ilia-cher replied:
I'd really really encourage not to do this though

@dzhulgakov  re. consistency - the only use case that does this is in distributed profiling code, and analyzing these async (start on one thread, end on another) ranges is a problematic, e.g. profiler wouldn't report much about them

so I'd say, unless this is absolutely needed, we don't use RecordFunction across thread boundaries

ilia-cher replied:
another use case is when code in TS interpreter calls wait, and then wakes up in another thread, in this case we get an async range too, but that's mostly unavoidable, unless you want to finish recording when code calls wait

ShijunK replied:
So, should we keep as it is? Today's "forked" function will be executed in another thread, has its own record function, which is unrelated to the original one (return immediately with future).

dzhulgakov replied:
Yeah, I guess we can keep it as is for now then. It won't attribute continuations to the same original invocation, but it seems that profiler is not too good at visualizing those anyway.

dzhulgakov reviewed (approved):
Please fix the clang-format before landing. Otherwise looks good.

ShijunK committed:
Expose run_async function on torch::jit::Method (#48607)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/48607

This change builds on top of
https://github.com/pytorch/pytorch/pull/46865

further exposing the async interface to `torch::jit::Method`.

added unit test for new `run_async`

Test Plan: `buck test caffe2/test/cpp/jit/...`

Reviewed By: dzhulgakov

Differential Revision: D25219726

fbshipit-source-id: f238e9765deed49c0b8384aaaae44e611f281476
